spec/SPEC-annex-A-filesystem-storage.md

```diff
@@ -16,11 +16,19 @@ Example:
 A build tool may persist additional metadata to that makes reference to the Artifact Dependency Graph (ADG).
 It should persist such metadata to a subdirectory of the directory to which the output artifact is being written of the form: ```${OMNIBOR_DIR}/metadata/${context}/```.  
 
-For metadata specific to a particular build tool ```${context}``` should be a name uniquely associated with the build tool.  Build tools should report their selection of ```${context}``` subdirectory name to the OmniBOR spec for inclusion in a list to preclude ```${context}``` collision.
+For metadata specific to a particular build tool ```${context}``` should be a name uniquely associated with the build tool.  For example: 
 
-Metadata persisted by multiple build tools in the same way should be documented in a specification for that metadata.  Such specs must include the ```${context}``` for that metadata.  Such specs should be reported to the OmniBOR spec for inclusion in a list to preclude ```${context}``` collision.
+- ```${OMNIBOR_DIR}/metadata/llvm```
+- ```${OMNIBOR_DIR}/metadata/clang```
+- ```${OMNIBOR_DIR}/metadata/go```
+- ```${OMNIBOR_DIR}/metadata/rustc```
+- ```${OMNIBOR_DIR}/metadata/gcc```
 
```

Comment:
Currently we generate the Artifact/Omnibor Input Manifest (IM) document under OMNIBOR_DIR. For example:
.omnibor/objects/gitoid_blob_sha1/66/f02ce601055296e27b55dd8dbac765ebd7812e

Instead of defining a different path/protocol to store the metadata, could we generate the metadata in the same location having the same basename as the IM, but with an extension, like .metadata or .md ? 
Some of the advantages of this approach are
- This is very simple to implement -- tools only need to create another file in parallel to the IM.
- This is also inline with how build tools like gcc or llvm create different output artifacts when required, with the same base name as that of the output artifact, but with different extensions that signifies the contents of the file. (For example, the --save-temps option when compiling foo.c into foo.o, generates foo.i, foo.s, foo.o and the -MD option generates foo.d). 
- The first line in this metadata file could indicate the 'context' or the name of the tool that generated this (eg., clang, lld, gcc, as, ld, rustc etc)
- It is trivial to establish the correlation between the IM and its metadata.

Reply:
This is an interesting idea.  How would we tell apart different metadata from different tools?  If I compiled something with clang, vs it being compiled with gcc.  How would we know which metadata was which?

Reply:
hi Bharathi, I like your proposal, which has many advantages as you listed.

I just want to share some thoughts on its potential issue in some scenarios.

For example, in RPM/DEB build, if you create such new hello.metadata in the same directory as hello, then this hello.metadata may mistakenly packaged into RPM/DEB.
In some Cisco build framework, some files in a directory are all signed, and if we create a new hello.metadata file in the same directory, the signing can fail, causing build failure.

However, these should be usually corner cases, and should be able to be solved easily.

Reply:
@yonhan3 Do you have ideas to solve these corner cases?

Reply:
> This is an interesting idea. How would we tell apart different metadata from different tools? If I compiled something with clang, vs it being compiled with gcc. How would we know which metadata was which? Is that even a real problem to consider?

The first line in the metadata file could indicate the tool name that created this metadata file (i.e, the context as in the current proposal). This would be similar to the header (e.g, gitoid_blob_sha1) we add to the IM to distinguish the sha types. 
Also, I'm assuming the command line options would be one of the metadata info to be captured. So we have an omnibor IM (foo) and its metadata(foo.metadata), so it becomes easier to understand how the artifact/omnibor IM itself was derived.

Reply:
> For example, in RPM/DEB build, if you create such new hello.metadata in the same directory as hello, then this hello.metadata may mistakenly packaged into RPM/DEB. 

@yonhan3  Are you saying it is not desirable for metadata files to be under OMNIBOR_DIR or it should not be in the same dir as the IM itself ? We could create hello.metadata under $OMNIBOR_DIR/metadata, which is not the same dir as the IM. I was mainly suggesting the naming of the metadata file as <OmniborIM>.metadata for easy correlation of the IM and its metadata.

Reply:
I might have misunderstood your proposal. I thought you directly save metadata file in the same directory of the generated files.
Using sha1.metadata as the name is good idea. We need to also consider that the same sha1 can have multiple file paths and multiple build commands, which has been observed many times.
For metadata, we should be flexible to support multiple alternatives: either a few large files to contain metadata for many artifacts, or many small files for each individual artifact respectively.